rspace/src/main/scala/coop/rchain/rspace/merger/MergingLogic.scala

```diff
@@ -0,0 +1,107 @@
+package coop.rchain.rspace.merger
+
+import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.trace.{Consume, Produce}
+
+import scala.Function.tupled
+
+object MergingLogic {
+
+  /** if target depends on source */
+  def depends(target: EventLogIndex, source: EventLogIndex): Boolean =
+    (producesActive(source) intersect target.producesConsumed).nonEmpty ||
+      (consumesActive(source) intersect target.consumesProduced).nonEmpty
+
+  /** if two event logs are conflicting */
+  def areConflicting(a: EventLogIndex, b: EventLogIndex): Boolean =
+    conflicts(a: EventLogIndex, b: EventLogIndex).nonEmpty
+
+  /** Channels conflicting between a pair of event logs */
+  def conflicts(a: EventLogIndex, b: EventLogIndex): Iterator[Blake2b256Hash] = {
+
+    /**
+      * Check #1
+      * If the same produce or consume is destroyed in COMM in both branches, this might be a race.
+      * All events created in event logs are unique, this match can be identified by comparing case classes.
+      *
+      * Produce is considered destroyed in COMM if it is not persistent and been consumed without peek.
+      * Consume is considered destroyed in COMM when it is not persistent.
+      */
+    val racesForSameIOEvent = {
+      def equal[A](as: Set[A], bs: Set[A]): Iterator[A] =
+        as.iterator
+          .flatMap(a => bs.iterator.map(b => (a, b)))
+          .filter(tupled((l, r) => l == r))
+          .map(_._1)
+
+      val consumeRaces =
+        equal[Consume](a.consumesProduced, b.consumesProduced).filterNot(_.persistent)
+      val produceRaces =
+        equal[Produce](a.producesConsumed, b.producesConsumed).filterNot(_.persistent)
+
+      consumeRaces.flatMap(_.channelsHashes) ++ produceRaces.map(_.channelsHash)
+    }
+
+    /**
+      * Check #2
+      * Events that are created inside branch and has not been destroyed in branch's COMMs
+      * can lead to potential COMM during merge.
+      */
+    val potentialCOMMs = {
+      // TODO analyze joins to make less conflicts. Now plain channel intersection treated as a conflict
+      def matchFound(consume: Consume, produce: Produce): Boolean =
+        consume.channelsHashes contains produce.channelsHash
```

Comment:
Now this is very restrictive - even if produce matches a single channel from consume, match is declared. Future work - find if there is a combination of produces that exactly matches consume, with pattern matching.